src/full/Agda/TypeChecking/Serialise/Instances/Abstract.hs

```diff
@@ -71,16 +75,41 @@ instance EmbPrj NameSpace where
 
   value = valueN NameSpace
 
-instance EmbPrj WhyInScope where
-  icod_ Defined       = icodeN' Defined
-  icod_ (Opened a b)  = icodeN 0 Opened a b
-  icod_ (Applied a b) = icodeN 1 Applied a b
+-- Andreas, 2025-10-23:
+-- We serialize the Ranges in WhyInScope
+-- so that we have then in the ClashingDefinition error.
+
+data WhyInScopeWithRange
+  = Defined'
+  | Opened'  SerialisedRange C.QName WhyInScope
+  | Applied' SerialisedRange C.QName WhyInScope
+
+instance EmbPrj WhyInScopeWithRange where
+  icod_ Defined'         = icodeN' Defined'
+  icod_ (Opened'  r a b) = icodeN 0 Opened'  r a b
+  icod_ (Applied' r a b) = icodeN 1 Applied' r a b
+
+  value = vcase \case
+    N0           -> valuN Defined'
+    (N4 0 r a b) -> valuN Opened'  r a b
+    (N4 1 r a b) -> valuN Applied' r a b
+    _            -> malformed
+
+toWhyInScopeWithRange :: WhyInScope -> WhyInScopeWithRange
+toWhyInScopeWithRange = \case
+  Defined     -> Defined'
+  Opened  x w -> Opened'  (SerialisedRange (getRange x)) x w
+  Applied x w -> Applied' (SerialisedRange (getRange x)) x w
+
+fromWhyInScopeWithRange :: WhyInScopeWithRange -> WhyInScope
+fromWhyInScopeWithRange = \case
+  Defined'                         -> Defined
+  Opened'  (SerialisedRange r) x w -> Opened  (setRange r x) w
+  Applied' (SerialisedRange r) x w -> Applied (setRange r x) w
 
-  value = vcase valu where
-    valu N0         = valuN Defined
-    valu (N3 0 a b) = valuN Opened a b
-    valu (N3 1 a b) = valuN Applied a b
-    valu _          = malformed
+instance EmbPrj WhyInScope where
+  icod_ = icod_ . toWhyInScopeWithRange
+  value = fromWhyInScopeWithRange <.> value
```

Comment:
Because the type-level programming in `icod_`, I could not directly fit in a `SerialisedRange` in its clauses.
So I first convert `WhyInScope` to `WhyInScopeWithRange` which is then serialized in the standard way.
But this feels very clumsy.
@AndrasKovacs : comment?

Reply:
Something like the following should work (I haven't yet checked):

```haskell
instance EmbPrj WhyInScope where
  icod_ Defined       = icodeN' Defined
  icod_ (Opened a b)  = icodeN 0 (\(_ :: SerialisedRange) -> Opened) (SerialisedRange (getRange a)) a b
  icod_ (Applied a b) = icodeN 1 (\(_ :: SerialisedRange) -> Applied) (SerialisedRange (getRange a)) a b

  value = vcase \case
    N0           -> valuN Defined
    (N4 0 r a b) -> valuN (\(SerialisedRange r) a b -> Opened (setRange r a) b) r a b
    (N4 1 r a b) -> valuN (\(SerialisedRange r) a b -> Applied (setRange r a) b) r a b
    _            -> malformed

Reply:
Thanks, that worked out of the box.